File: core4/leetcode/Arrays/P697/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
       
        unordered_map<int,vector<int>> mp;  // int: number → vector:{count, first_index, last_index}
       
        for(int index=0 ; index < nums.size(); index++){
            int x = nums[index];
            if(mp.count(x)){
                mp[x][0]++;        // increment count
                mp[x][2] = index;  // update last index
            } else{
                mp[x] = {1, index, index}; // first occurrence
            }
        }
       
       
        int degree = 0;
        for(const auto& entry  : mp){
            if (entry.second[0]  > degree){
                degree = entry.second[0];
            }
        }
       
        int minLength = nums.size();
        for(const auto& entry  : mp){
            if(entry.second[0] == degree){
                int pot_minLength = entry.second[2] - entry.second[1] + 1;
                if(pot_minLength < minLength){
                    minLength = pot_minLength;
                }
            }
        }
       
        return minLength;
    }
};


int main()
{
    vector<int> nums = {1,2,2,3,1,2,3,2,4,4,6};
    Solution object;
    cout << "Smallest subarray length with the array degree: " << object.findShortestSubArray(nums);
   

}

File: core4/leetcode/Arrays/P697/main.py
class Solution:
    def findShortestSubArray(self,num):
        if not num:
            return 0
       
        mp={}  # number → [count, first_index, last_index]
       
        for index in range(len(num)):
            x = num[index]
            if x in mp:
                mp[x][0] +=1
                lastindex = index    # increment count
                mp[x][2] = lastindex # update last index
            else:
                mp[x] = [1,index,index]
               
        degree = 0  # Keeps track of the maximum frequency
        for x in mp:
            if mp[x][0] > degree:
                degree = mp[x][0]
       
        min_length = len(num)
        for x in mp:
            if mp[x][0] == degree:
                pot_min_length = mp[x][2] - mp[x][1] + 1
                if pot_min_length < min_length:
                    min_length = pot_min_length
                   
        return min_length
                   
       
       
       
       
num = [1,2,2,3,1]
sol = Solution()
print("The minimum array length is: ", sol.findShortestSubArray(num))  

